src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs

```diff
@@ -121,11 +124,104 @@ protected override void RemoveItem(int index)
 
             base.RemoveItem(index);
 
+            if (!_skipRaisingEvents)
+            {
+                OnCountPropertyChanged();
+                OnIndexerPropertyChanged();
+                OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItem, index);
+            }
+        }
+
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a count of items is removed from the list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void RemoveItemsRange(int index, int count)
+        {
+            CheckReentrancy();
+
+            int length = this.Count;
+            if (index < this.Count - count)
+            {
+                length = index + count;
+            }
+
+            T[] oldItems = new T[Items.Count];
+            if (Items is List<T> list)
+            {
+                list.CopyTo(oldItems);
+            }
+            else
+            {
+                for (int i = 0; i < oldItems.Length; i++)
+                {
+                    oldItems[i] = this[i];
+                }
+            }
+
+            bool ignore = _skipRaisingEvents;
+            if (!ignore)
+            {
+                _skipRaisingEvents = true;
+            }
+
+            try
+            {
+                base.RemoveItemsRange(index, count);
+            }
+            finally
+            {
+                if (!ignore)
+                {
+                    _skipRaisingEvents = false;
+                }
+            }
+            
+            if (count > 0 && !_skipRaisingEvents)
+            {
+                T[] removedItems = new T[count];
+                for (int i = 0; i < count; i++)
+                {
+                    removedItems[i] = oldItems[index + i];
+                }
+
+                OnCountPropertyChanged();
+                OnIndexerPropertyChanged();
+                OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItems, index);
+            }
+        }
+
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a collection of items is added to list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void ReplaceItemsRange(int index, int count, IEnumerable<T> collection)
+        {
+            CheckReentrancy();
+
+            _skipRaisingEvents = true;
+
+            IList itemsToReplace = new List<T>();
```

Comment:
I still think we should copy the removed items to an array, like we do with `removedItems` in `RemoveItemsRange`, but if we're going to use a `List<T>` here for some reason, it'd be worth keeping `itemsToReplace` typed as `List<T>` instead of `IList`. That way, the `Add` calls won't have to go through interface dispatch.

```suggestion
            List<T> itemsToReplace = new List<T>();
```

You'll still be able to pass `itemsToReplace` to `OnCollectionChanged` as-is without any changes because `List<T>` implements `IList`.

Reply:
I took another look at converting this to an array and was able to get it working. Thanks!